smithy-utils/src/main/java/software/amazon/smithy/utils/CodeInterceptor.java

```diff
@@ -0,0 +1,191 @@
+ * the {@link AbstractCodeWriter} they are given when called.
+ *
+ * <p>Unless you need to intercept previously written content and change it,
+ * it's best to implement the {@link Appender} or {@link Prepender} interfaces
+ * since they take care of properly writing previously written content to
+ * the section (for example, only writing if it's non-empty, and using
+ * writeInlineWithNoFormatting to avoid unintentional interpolation).
+ *
+ * @param <S> Type of CodeSection to intercept.
+ * @param <W> Type of CodeWriter to expect.
+ */
+public
+interface CodeInterceptor<S extends CodeSection, W extends AbstractCodeWriter<W>> {
+
+    /**
+     * Get the strongly typed {@link CodeSection} this interceptor is used
+     * to intercept.
+     *
+     * @return The code section to intercept.
+     */
+    Class<S> sectionType();
+
+    /**
+     * Checks if the given section is filtered by this interceptor or not.
+     *
+     * <p>In some cases {@link #sectionType()} might allow filtering a wider
+     * array of types than what is actually filtered by an interceptor. The
+     * most common example of this is intercepting any type of CodeSection
+     * and only filtering based on the result of {@link CodeSection#sectionName()}.
+     *
+     * <p>This method will return {@code true} by default, meaning that any
+     * type of {@link CodeSection} that is an instance of the class returned
+     * from {@link #sectionType()} will be intercepted.
+     *
+     * @param section Section to test if this interceptor is relevant.
+     * @return Returns true if the section is intercepted or not.
+     */
+    default boolean isIntercepted(S section) {
+        return true;
+    }
+
+    /**
+     * Intercepts an {@link AbstractCodeWriter} section.
+     *
+     * @param writer       Writer used to write content. If no write calls are
+     *                     made, any intercepted text is lost.
+     * @param previousText The previous text that was written. This text
+     *                     needs to be written again in order for it to
+     *                     be kept in the section.
+     * @param section      The strongly typed section value.
+     */
+    void write(W writer, String previousText, S section);
+
+    /**
+     * Provides a more concise way of creating anonymous {@link Appender}s.
+     *
+     * <p>This method does not support custom filters on matched CodeSections. That
+     * functionality must be implemented by directly creating an Appender class.
+     *
+     * @param type The type of section to intercept.
+     * @param appender A BiConsumer that takes the writer and section and is expected to make write calls.
+     * @param <S> The type of section being intercepted.
+     * @param <W> The type of writer to use.
+     * @return Returns the created Appender.
+     */
+    static <S extends CodeSection, W extends AbstractCodeWriter<W>> CodeInterceptor<S, W> appender(
```

Comment:
Any reason a concise version of `prepender` doesn't exist?

Reply:
I thought about it, but I think it'll be significantly more rare than appending, and you can still create a prepender manually if needed. We can always add it later was my thought.